### DISCUSSION TITLE
Atoms in atom vs atomFamily

### DISCUSSION THREAD

**Discussion by oferitz** [pmndrs/jotai#2295] [Q&A] [answered]
I'm in the process of transitioning from Recoil and need to determine the best approach for migration. My specific scenario involves an application where users can create workflows by dragging actions onto a canvas and connecting them. Each action possesses its own state, typically resembling a form with inputs that dictate a set of instructions for execution based on the input values.

In Recoil, each action's state is stored in a Recoil atom, which is part of an atom family. The primary objective is to selectively rerender only the current action when its inputs change, without affecting the entire canvas or other actions. When the user decides to save the workflow, we consolidate the states of all actions into a single payload object using a selectorFamily.

Now, as I embark on the migration process with Jotai, I'm deliberating between utilizing atoms in atom or atomFamily. Which approach would be more suitable for implementing this functionality with Jotai?

**Comment by dai-shi** [comment]
tbh, I'm not sure if I understand your requirement entirely. Maybe some code snippet might help.
atomFamily is originally developed based on the request from Recoil community.
So, in general, it would be helpful for migrating from Recoil.
Though, the API is not the same and migration isn't one-to-one. In that sanse, converting to atom-in-atom patten might be reasonable.
There is nothing wrong with atomFamily at all. It's recommended if the set of params is finite, or if you take the responsibility of removing unused params.

**Reply by oferitz** [reply]
@dai-shi Thank you for responding. My fundamental question is this: When implemented as an atomFamily, can we guarantee that updating an atom within the family will only trigger a re-render in the component utilizing that specific atom family member?

**Reply by dai-shi** [reply, marked as answer]
> can we guarantee that updating an atom within the family will only trigger a re-render

Sure. I recommend reading a simplified version of atomFamily implementation as blow:

https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035

```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```

Basically, it's a Map, and there's nothing special tied to Jotai core.
